### PR TITLE
Certora

### PR DESCRIPTION
**Pull request by oldchili**

### CODE REVIEW

File: test/mocks/GemMock.sol [hunk 1]
```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+pragma solidity ^0.8.13;
+
+contract GemMock {
```
Comment (lucas-manuel) [inline test/mocks/GemMock.sol:5]: We can just import MockERC20 for this: https://github.com/marsfoundation/erc20-helpers/blob/9cf80a0d21692a4442c38bf2cb76c866ee3b7d85/src/MockERC20.sol

File: certora/harness/ArrangerConduitHarness.sol [hunk 2]
```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.13;
+
+import {ArrangerConduit} from "src/ArrangerConduit.sol";
```
Comment (lucas-manuel) [inline certora/harness/ArrangerConduitHarness.sol:4]: { ArrangerConduit }

**Comment by oldchili** [comment]
Going to do a few structural changes that @sunbreak1211 suggested, so converting back to draft for a short while.

File: test/mocks/GemMock.sol (hunk 1 above)
Comment (oldchili) [reply test/mocks/GemMock.sol]: Done.

File: certora/harness/ArrangerConduitHarness.sol (hunk 2 above)
Comment (oldchili) [reply certora/harness/ArrangerConduitHarness.sol]: Eventually removed this file.

File: certora/ArrangerConduit.conf [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{
+    "files": [
+        "src/ArrangerConduit.sol",
+        "lib/dss-allocator/src/AllocatorRoles.sol",
+        "lib/dss-allocator/src/AllocatorRegistry.sol",
+        "lib/mock-erc20/src/MockERC20.sol",
+        "certora/Auxiliar.sol"
+
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.conf:8]:
```suggestion
```

File: certora/ArrangerConduit.conf [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    ],
+    "link": [
+        "ArrangerConduit:roles=AllocatorRoles",
+        "ArrangerConduit:registry=AllocatorRegistry",
+    ],
+    "optimistic_loop": true,
+    "rule_sanity": "basic",
+    "solc_optimize_map": {
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.conf:16]: It's missing to specify a general solc version or a map of them.

File: certora/ArrangerConduit.spec [hunk 5]
```diff
@@ -0,0 +1,710 @@
...
+        address    asset;  // 20 bytes
+        bytes32    ilk;  // 32 bytes
+        uint256    amountRequested; // 32 bytes
+        uint256    amountFilled;  // 32 bytes
+        string     info;
+    }
+  * The storage is 32-bytes aligned. If each field gets 32-bytes, then the offset
+  * would be 32 * 5 = 160. If the `status` and `asset` fields are packed together,
+  * then the offset would be 32 * 4 = 128.
+  * Trial and error shows the offset is 128.
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:83]: Not sure we really need to leave this as a comment as we know that status and asset will be packed in one storage slot.

File: certora/ArrangerConduit.spec [hunk 6]
```diff
@@ -0,0 +1,710 @@
...
+    setBroker(e, usr, asset, valid);
+
+    bool isBrokerUsrAfter = isBroker(usr, asset);
+    bool isBrokerOtherUsrAfter = isBroker(otherUsr, asset);
+    bool isBrokerOtherAssetAfter = isBroker(usr, otherAsset);
+
+    assert isBrokerUsrAfter == valid, "setBroker did not set broker to valid";
+    assert isBrokerOtherUsrAfter == isBrokerOtherUsrBefore, "setBroker did not keep unchanged the rest of brokers for the same asset";
+    assert isBrokerOtherAssetAfter == isBrokerOtherAssetBefore, "setBroker did not keep unchanged the rest of brokers for other assets";
+}
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:259]: I solve this cases using the following:

```suggestion
// Verify correct storage changes for non reverting setBroker
rule setBroker(address usr, address asset, bool valid) {
    env e;

    address otherUsr;
    address otherAsset;
    require otherUsr != usr || otherAsset != asset;
    
    bool isBrokerOtherBefore = isBroker(otherUsr, otherAsset);
    
    setBroker(e, usr, asset, valid);

    bool isBrokerUsrAfter = isBroker(usr, asset);
    bool isBrokerOtherAfter = isBroker(otherUsr, otherAsset);

    assert isBrokerUsrAfter == valid, "setBroker did not set brokers[usr][asset] to valid";
    assert isBrokerOtherAfter == isBrokerOtherBefore, "setBroker did not keep unchanged the rest of brokers[x][y]";
}
```
Not only because it is shorter, but also because the way done now it is excluding the case where `isBroker(otherUser, otherAsset)`.

File: certora/ArrangerConduit.spec [hunk 7]
```diff
@@ -0,0 +1,710 @@
...
+
+    require asset == gem;
+
+    address buffer = registry.buffers(ilk);
+    require buffer != currentContract;
+
+    mathint balanceOfBufferBefore = gem.balanceOf(buffer);
+    mathint balanceOfConduitBefore = gem.balanceOf(currentContract);
+    mathint depositsBefore = deposits(asset, ilk);
+    mathint totalDepositsBefore = totalDeposits(asset);
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:289]: Following the pattern that we try to check every storage slot, we should be checking deposits for other assets or ilks and totalDeposits for other assets. Important to notice here is that we should be decoupling the "other" variable used for asset in deposits than the one using for total (if you follow the way I handle these cases mentioned above).
So I picture something like this:
```
address otherIlk;
address otherAsset;
require otherIlk != ilk || otherAsset != asset;

address otherAsset2;
require otherAsset2 != asset;

mathint depositsOtherBefore = deposits(otherAsset, otherIlk);
mathint totalDepositsOtherBefore = totalDeposits(otherAsset2);
```
This way we can ensure that all possibilities for both mappings can be checked.

File: certora/ArrangerConduit.spec [hunk 8]
```diff
@@ -0,0 +1,710 @@
...
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = !canCall;
+    bool revert3 = balanceOfBuffer < to_mathint(amount);
+    bool revert4 = allowanceBuffer < to_mathint(amount);
+    bool revert5 = deposits + amount > max_uint256;
+    bool revert6 = totalDeposits + amount > max_uint256;
+    bool revert7 = buffer == 0;
+
+    assert lastReverted <=> revert1 || revert2 || revert3 ||
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:331]: Couldn't actually run it yet due to the submodule issue, but weird we are not having a revert condition about overflow on gem balance for the conduit address.

File: certora/ArrangerConduit.spec [hunk 9]
```diff
@@ -0,0 +1,710 @@
...
+    require buffer != currentContract;
+
+    bool canCall = roles.canCall(ilk, e.msg.sender, currentContract, to_bytes4(0xa6fb97d1)); // withdraw(bytes32,address,uint256)
+    mathint balanceOfConduit = gem.balanceOf(currentContract);
+    mathint withdrawals = withdrawals(asset, ilk);
+    mathint totalWithdrawals = totalWithdrawals(asset);
+    mathint withdrawableFunds = withdrawableFunds(asset, ilk);
+    mathint totalWithdrawableFunds = totalWithdrawableFunds(asset);
+
+    mathint amount = min(maxAmount, maxWithdraw(ilk, asset));
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:389]:
```suggestion
    mathint amount = min(maxAmount, maxWithdraw(ilk, asset));
```

File: certora/ArrangerConduit.spec [hunk 10]
```diff
@@ -0,0 +1,710 @@
...
+    mathint balanceOfConduit = gem.balanceOf(currentContract);
+    mathint withdrawals = withdrawals(asset, ilk);
+    mathint totalWithdrawals = totalWithdrawals(asset);
+    mathint withdrawableFunds = withdrawableFunds(asset, ilk);
+    mathint totalWithdrawableFunds = totalWithdrawableFunds(asset);
+
+    mathint amount = min(maxAmount, maxWithdraw(ilk, asset)); 
+    withdraw@withrevert(e, ilk, asset, maxAmount);
+
+    bool revert1 = e.msg.value > 0;
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:392]: Same comment, weird we are not having a revert on balance of the destination.

Comment (sunbreak1211) [reply certora/ArrangerConduit.spec:392]: On the other hand, I have a personal preference to try to preserve the order of the reverts according to how they can happen in the code. Sometimes it is a bit difficult though as some revert rule might group different revert conditions, and also a bit annoying if you discovered one is missing in the beginning.

File: certora/ArrangerConduit.conf [hunk 11]
```diff
@@ -0,0 +1,28 @@
+{
+    "files": [
+        "src/ArrangerConduit.sol",
+        "lib/dss-allocator/src/AllocatorRoles.sol",
+        "lib/dss-allocator/src/AllocatorRegistry.sol",
+        "lib/mock-erc20/src/MockERC20.sol",
```
[the remark this answers is not in the document]
Comment (oldchili) [reply certora/ArrangerConduit.conf:6]: Ah I see it was removed at some point an I still have it locally as a leftover.
Will reinstall it as that's what @lucas-manuel suggested.

[the remark this answers is not in the document]
Comment (telome) [reply certora/ArrangerConduit.conf:6]: Looks like the file is in `lib/erc20-helpers/src/MockERC20.sol`

File: certora/ArrangerConduit.spec [hunk 12]
```diff
@@ -0,0 +1,710 @@
...
+    require asset == gem;
+
+    address buffer = registry.buffers(ilk);
+    require buffer != currentContract;
+
+    bool canCall = roles.canCall(ilk, e.msg.sender, currentContract, to_bytes4(0xd954863c)); // deposit(bytes32,address,uint256)
+    mathint balanceOfBuffer = gem.balanceOf(buffer);
+    mathint allowanceBuffer = gem.allowance(buffer, currentContract);
+    mathint deposits = deposits(asset, ilk);
+    mathint totalDeposits= totalDeposits(asset);
```
Comment (telome) [inline certora/ArrangerConduit.spec:319]:
```suggestion
    mathint totalDeposits = totalDeposits(asset);
```

File: certora/ArrangerConduit.spec (hunk 8 above)
Comment (telome) [reply certora/ArrangerConduit.spec:331]: Yeah that seems to be because the erc20 mock cannot overflow during transfers:

image


I tested with  the `unchecked` removed from `_transfer()` and can confirm that the extra revert condition you mention is indeed required in that case.

File: certora/ArrangerConduit.spec [hunk 13]
```diff
@@ -0,0 +1,710 @@
...
+    require balanceOfBufferBefore + balanceOfConduitBefore <= max_uint256;
+
+    mathint amount = min(maxAmount, maxWithdraw(ilk, asset)); 
+    withdraw(e, ilk, asset, maxAmount);
+
+    mathint balanceOfBufferAfter = gem.balanceOf(buffer);
+    mathint balanceOfConduitAfter = gem.balanceOf(currentContract);
+    mathint withdrawableFundsAfter = withdrawableFunds(asset, ilk);
+    mathint totalWithdrawableFundsAfter = totalWithdrawableFunds(asset);
+    mathint withdrawalsAfter= withdrawals(asset, ilk);
```
Comment (telome) [inline certora/ArrangerConduit.spec:362]:
```suggestion
    mathint withdrawalsAfter = withdrawals(asset, ilk);
```

File: certora/ArrangerConduit.spec [hunk 14]
```diff
@@ -0,0 +1,710 @@
...
+    mathint totalWithdrawalsAfter = totalWithdrawals(asset);
+
+    assert balanceOfBufferAfter == balanceOfBufferBefore + amount, "balance of buffer did not increase by amount";
+    assert balanceOfConduitAfter == balanceOfConduitBefore - amount, "balance of conduit did not decrease by amount";
+    assert withdrawableFundsAfter == withdrawableFundsBefore - amount, "withdrawableFunds did not decrease by amount";
+    assert totalWithdrawableFundsAfter == totalWithdrawableFundsBefore- amount, "totalWithdrawableFunds did not decrease by amount";
```
Comment (telome) [inline certora/ArrangerConduit.spec:368]:
```suggestion
    assert totalWithdrawableFundsAfter == totalWithdrawableFundsBefore - amount, "totalWithdrawableFunds did not decrease by amount";
```

File: certora/ArrangerConduit.spec (hunk 8 above)
Comment (sunbreak1211) [reply certora/ArrangerConduit.spec:330]: Ok yeah 👍

File: certora/ArrangerConduit.conf (hunk 4 above)
Comment (oldchili) [reply certora/ArrangerConduit.conf:16]: Done

File: certora/ArrangerConduit.spec (hunk 5 above)
Comment (oldchili) [reply certora/ArrangerConduit.spec]: Done

File: certora/ArrangerConduit.conf (hunk 11 above)
[the remark this answers is not in the document]
Comment (oldchili) [reply certora/ArrangerConduit.conf]: Yeah nice, moved to used the updated path.

File: certora/ArrangerConduit.spec (hunk 7 above)
Comment (oldchili) [reply certora/ArrangerConduit.spec:282]: Ok, done more or less that, also in other places.

File: certora/ArrangerConduit.spec (hunk 10 above)
Comment (oldchili) [reply certora/ArrangerConduit.spec:429]: nvm I just did it.

File: certora/ArrangerConduit.spec [hunk 15]
```diff
@@ -0,0 +1,805 @@
...
+    require otherIlk != ilk || otherAsset != asset;
+    mathint withdrawableFundsOtherBefore = withdrawableFunds(otherAsset, otherIlk);
+
+    address otherAsset2;
+    require otherAsset2 != asset;
+    mathint totalWithdrawableFundsOtherBefore = totalWithdrawableFunds(otherAsset2);
+
+    bytes32 otherIlk2;
+    address otherAsset3;
+    require otherIlk2 != ilk || otherAsset3 != asset;
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:372]: This is not wrong at all, but you could use the same `otherIlk` and `otherAsset` here, as they are tight to the same grouped key. As you just need to find a condition where this doesn't pass, it doesn't really matter if for each run both mappings are checked on the same key.
The problem to use `otherAsset` where you are using `otherAsset2` is that `otherAsset` could be equal to `asset` due to the `||` condition, and there you will have an unintended violation.
So basically, summarizing, you just need `otherIlk`, `otherAsset` and `otherAsset2` for this rule.

File: certora/ArrangerConduit.spec [hunk 16]
```diff
@@ -0,0 +1,805 @@
...
+    mathint withdrawalsOtherBefore = withdrawals(otherAsset3, otherIlk2);
+
+    address otherAsset4;
+    require otherAsset4 != asset;
+    mathint totalWithdrawalsOtherBefore = totalWithdrawals(otherAsset4);
+
+    require balanceOfBufferBefore + balanceOfConduitBefore <= max_uint256;
+
+    mathint amount = min(maxAmount, maxWithdraw(ilk, asset));
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:381]: nit but I feel looks more correct:
```suggestion
    mathint amount = min(maxAmount, withdrawableFundsBefore); 
```

File: certora/ArrangerConduit.spec [hunk 17]
```diff
@@ -0,0 +1,805 @@
...
+    assert balanceOfConduitAfter == balanceOfConduitBefore - amount, "balance of conduit did not decrease by amount";
+    assert withdrawableFundsAfter == withdrawableFundsBefore - amount, "withdrawableFunds did not decrease by amount";
+    assert totalWithdrawableFundsAfter == totalWithdrawableFundsBefore - amount, "totalWithdrawableFunds did not decrease by amount";
+    assert withdrawalsAfter == withdrawalsBefore + amount, "withdrawals did not increase by amount";
+    assert totalWithdrawalsAfter == totalWithdrawalsBefore + amount, "totalWithdrawals did not increase by amount";
+
+    assert withdrawableFundsOtherAfter == withdrawableFundsOtherBefore, "other withdrawable funds changed unexpectedly";
+    assert totalWithdrawableFundsOtherAfter == totalWithdrawableFundsOtherBefore, "other total withdrawable funds changed unexpectedly";
+    assert withdrawalsOtherAfter == withdrawalsOtherBefore, "other withdrawals changed unexpectedly";
+    assert totalWithdrawalsOtherAfter == totalWithdrawalsOtherAfter, "other total withdrawals changed unexpectedly";
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:406]:
```suggestion
    assert totalWithdrawalsOtherAfter == totalWithdrawalsOtherBefore, "other total withdrawals changed unexpectedly";
```

File: certora/ArrangerConduit.spec [hunk 18]
```diff
@@ -0,0 +1,805 @@
...
+    require buffer != currentContract;
+
+    bool canCall = roles.canCall(ilk, e.msg.sender, currentContract, to_bytes4(0xa6fb97d1)); // withdraw(bytes32,address,uint256)
+    mathint balanceOfConduit = gem.balanceOf(currentContract);
+    mathint withdrawals = withdrawals(asset, ilk);
+    mathint totalWithdrawals = totalWithdrawals(asset);
+    mathint withdrawableFunds = withdrawableFunds(asset, ilk);
+    mathint totalWithdrawableFunds = totalWithdrawableFunds(asset);
+
+    mathint amount = min(maxAmount, maxWithdraw(ilk, asset));
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:426]: Same nit:
```suggestion
    mathint amount = min(maxAmount, withdrawableFunds);
```

File: certora/ArrangerConduit.spec [hunk 19]
```diff
@@ -0,0 +1,805 @@
...
+    withdraw@withrevert(e, ilk, asset, maxAmount);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = !canCall;
+    bool revert3 = withdrawableFunds < amount;
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:431]: This shouldn't be a revert condition as amount is always limiting amount.

File: certora/ArrangerConduit.spec [hunk 20]
```diff
@@ -0,0 +1,805 @@
...
+    bool revert4 = totalWithdrawableFunds < amount;
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:432]: This one is fine, but we might consider make an invariant proving that `totalWithdrawableFunds` equals to the sum of  `withdrawableFunds` (or at least >= than anyone). Then if we require the invariant that revert is not actually necessary.

File: certora/ArrangerConduit.spec [hunk 21]
```diff
@@ -0,0 +1,805 @@
...
+    require otherAsset2 != asset;
+    mathint totalRequestedFundsOtherBefore = totalRequestedFunds(otherAsset2); 
+
+    IArrangerConduit.FundRequest requestBefore = getFundRequest(anyIndex);
+    bytes32 infoHashBefore = aux.hashString(requestBefore.info);
+
+    requestFunds(e, ilk, asset, amount, info);
+
+    mathint requestedFundsAfter = requestedFunds(asset, ilk);
+    mathint totalRequestedFundsAfter= totalRequestedFunds(asset);
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:469]:
```suggestion
    mathint totalRequestedFundsAfter = totalRequestedFunds(asset);
```

File: certora/ArrangerConduit.spec [hunk 22]
```diff
@@ -0,0 +1,805 @@
...
+    assert lastReverted <=> revert1 || revert2 || revert3 ||
+                            revert4 || revert5 || revert6 ||
+                            revert7 || revert8, "Revert rules failed";
+}
+
+// Verify correct storage changes for non reverting requestFunds
+rule requestFunds(bytes32 ilk, address asset, uint256 amount, string info) {
+    env e;
+
+    require asset == gem;
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:447]:
```suggestion
```
Not needed as there isn't any external call to it to evaluate.

File: certora/ArrangerConduit.spec [hunk 23]
```diff
@@ -0,0 +1,805 @@
...
+    assert requestedFundsOtherAfter == requestedFundsOtherBefore, "other requested funds changed unexpectedly";
+    assert totalRequestedFundsOtherAfter == totalRequestedFundsOtherBefore, "other total requested funds changed unexpectedly";
+}
+
+// Verify revert rules on requestFunds
+rule requestFunds_revert(bytes32 ilk, address asset, uint256 amount, string info) {
+    env e;
+
+    require asset == gem;
+
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:509]:
```suggestion
```

File: certora/ArrangerConduit.spec [hunk 24]
```diff
@@ -0,0 +1,805 @@
...
+    assert anyIndex != fundRequestId => requestAfter.status == requestBefore.status, "cancelFundRequest on another index changed status";
+    assert requestAfter.asset == requestBefore.asset
+        && requestAfter.ilk == requestBefore.ilk
+        && requestAfter.amountRequested == requestBefore.amountRequested
+        && requestAfter.amountFilled == requestBefore.amountFilled
+        && infoHashAfter == infoHashBefore,
+        "other request params not as before";
+
+    assert anyIndex == fundRequestId => requestedFundsOtherAfter == requestedFundsOtherBefore, "other requested funds changed unexpectedly";
+    assert anyIndex == fundRequestId => totalRequestedFundsOtherAfter == totalRequestedFundsOtherBefore, "other total requested funds changed unexpectedly";
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:573]: It's a bit weird how this thing ended up. Maybe is easier to think if we would have a `requestBefore/After` using the `fundRequestId` and then a `requestOtherBefore/After` using `anyIndex` which should be `!= fundRequestId`.
That way it would be more inline how we treat other storages.

File: certora/ArrangerConduit.spec [hunk 25]
```diff
@@ -0,0 +1,805 @@
...
+
+// TODO: figure out why this is still not working
+// Verify revert rules on cancelFundRequest
+rule cancelFundRequest_revert(uint256 fundRequestId) {
+    env e;
+
+    IArrangerConduit.FundRequest request = getFundRequest(fundRequestId);
+
+    require request.asset == gem;
+
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:584]:
```suggestion
```

File: certora/ArrangerConduit.spec [hunk 26]
```diff
@@ -0,0 +1,805 @@
...
+                            revert4 || revert5, "Revert rules failed";
+
+}
+
+// Verify correct storage changes for non reverting requestFunds
+rule drawFunds(address asset, address destination, uint256 amount) {
+    env e;
+
+    require asset == gem;
+    require currentContract != destination;
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:607]: This is a bit of personal taste, but I prefer to put requires when we are sure some address can't be other. But in this case is not even a manageable address, is one being passed by parameter, so I'd probably avoid this require and I'd check the assert for both cases.

File: certora/ArrangerConduit.spec [hunk 27]
```diff
@@ -0,0 +1,805 @@
...
+    address arranger = arranger();
+
+    drawFunds@withrevert(e, asset, destination, amount);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = arranger != e.msg.sender;
+    bool revert3 = totalWithdrawableFunds > balanceOfConduit;
+    bool revert4 = to_mathint(amount) > balanceOfConduit - totalWithdrawableFunds;
+    bool revert5 = !isBroker;
+    bool revert6 = balanceOfConduit < to_mathint(amount);
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:643]: This should be covered by `revert4`.

File: certora/ArrangerConduit.spec [hunk 28]
```diff
@@ -0,0 +1,805 @@
...
+        && requestAfter.amountRequested == requestBefore.amountRequested
+        && requestAfter.amountFilled == requestBefore.amountFilled
+        && infoHashAfter == infoHashBefore,
+        "other request params not as before";
+
+    assert anyIndex == fundRequestId => requestedFundsOtherAfter == requestedFundsOtherBefore, "other requested funds changed unexpectedly";
+    assert anyIndex == fundRequestId => totalRequestedFundsOtherAfter == totalRequestedFundsOtherBefore, "other total requested funds changed unexpectedly";
+}
+
+// TODO: figure out why this is still not working
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:576]:
```suggestion
```

File: certora/ArrangerConduit.spec [hunk 29]
```diff
@@ -0,0 +1,805 @@
...
+    bytes32 otherIlk;
+    address otherAsset;
+    require otherIlk != requestBefore.ilk || otherAsset != requestBefore.asset;
+    mathint requestedFundsOtherBefore = requestedFunds(otherAsset, otherIlk);
+
+    address otherAsset2;
+    require otherAsset2 != requestBefore.asset;
+    mathint totalRequestedFundsOtherBefore = totalRequestedFunds(otherAsset2);
+
+    bytes32 otherIlk2;
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:675]: Same here, with `otherIlk`, `otherAsset` and `otherAsset2` should be enough for all the variables.

File: certora/ArrangerConduit.spec (hunk 17 above)
Comment (oldchili) [reply certora/ArrangerConduit.spec]: done

File: certora/ArrangerConduit.spec (hunk 29 above)
Comment (oldchili) [reply certora/ArrangerConduit.spec]: Done

File: certora/ArrangerConduit.spec (hunk 27 above)
Comment (oldchili) [reply certora/ArrangerConduit.spec]: right, removed

File: certora/ArrangerConduit.spec (hunk 26 above)
Comment (oldchili) [reply certora/ArrangerConduit.spec]: Done

File: certora/ArrangerConduit.spec [hunk 30]
```diff
@@ -602,16 +601,18 @@ rule drawFunds(address asset, address destination, uint256 amount) {
     mathint balanceOfConduitAfter = gem.balanceOf(currentContract);
     mathint balanceOfDestinationAfter = gem.balanceOf(destination);
 
-    assert balanceOfConduitAfter == balanceOfConduitBefore - amount, "balance of conduit did not decrease by amount";
-    assert balanceOfDestinationAfter == balanceOfDestinationBefore + amount, "balance of destination did not increase by amount";
+    assert currentContract != destination => balanceOfConduitAfter == balanceOfConduitBefore - amount, "balance of conduit did not decrease by amount";
+    assert currentContract == destination => balanceOfConduitAfter == balanceOfConduitBefore, "balance of conduit changed";
+    assert currentContract != destination => balanceOfDestinationAfter == balanceOfDestinationBefore + amount, "balance of destination did not increase by amount";
+    assert currentContract == destination => balanceOfDestinationAfter == balanceOfDestinationBefore, "balance of destination changed";
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:607]: If `currentContract == destination`, `balanceOfConduit == balanceOfDestination`, then you just need to check it once.

File: certora/ArrangerConduit.spec [hunk 31]
```diff
@@ -0,0 +1,784 @@
...
+    assert totalWithdrawableFundsAfter == totalWithdrawableFundsBefore - amount, "totalWithdrawableFunds did not decrease by amount";
+    assert withdrawalsAfter == withdrawalsBefore + amount, "withdrawals did not increase by amount";
+    assert totalWithdrawalsAfter == totalWithdrawalsBefore + amount, "totalWithdrawals did not increase by amount";
+
+    assert withdrawableFundsOtherAfter == withdrawableFundsOtherBefore, "other withdrawable funds changed unexpectedly";
+    assert totalWithdrawableFundsOtherAfter == totalWithdrawableFundsOtherBefore, "other total withdrawable funds changed unexpectedly";
+    assert withdrawalsOtherAfter == withdrawalsOtherBefore, "other withdrawals changed unexpectedly";
+    assert totalWithdrawalsOtherAfter == totalWithdrawalsOtherBefore, "other total withdrawals changed unexpectedly";
+}
+
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:401]:
```suggestion
```

File: certora/ArrangerConduit.spec (hunk 24 above)
Comment (oldchili) [reply certora/ArrangerConduit.spec:562]: Implemented here - 
https://github.com/makerdao/dss-conduits/pull/18

These rules are quite heavy on the eye, we can consider separating to several ones.

File: certora/ArrangerConduit.spec (hunk 20 above)
Comment (oldchili) [reply certora/ArrangerConduit.spec]: The invariant is basically covered in `changeTogether`.
I don't see an advantage for requiring the invariant vs having the revert condition.

File: certora/ArrangerConduit.spec (hunk 30 above)
Comment (oldchili) [reply certora/ArrangerConduit.spec]: ok, done.

File: certora/ArrangerConduit.spec [hunk 32]
```diff
@@ -0,0 +1,860 @@
...
+    // other (asset,ilk) for actual request Id before
+    mathint requestedFundsOtherAfter = requestedFunds(otherAsset, otherIlk);
+    mathint totalRequestedFundsOtherAfter = totalRequestedFunds(otherAsset2); 
+
+    // other request Id after
+    IArrangerConduit.FundRequest requestOtherAfter = getFundRequest(otherIndex);
+    bytes32 infoHashOtherAfter = aux.hashString(requestOtherBefore.info);
+
+    // general asserts
+    assert numRequestsAfter == numRequestsBefore, "num requests changed";
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:587]: Probably this one should be evaluated in storageChanges rule and just in a specific rule when it actually changes.

File: certora/ArrangerConduit.spec [hunk 33]
```diff
@@ -0,0 +1,860 @@
...
+    mathint withdrawalsDiff = withdrawals(anyAsset, anyIlk) - withdrawalsBefore;
+    mathint totalDepositsDiff = totalDeposits(anyAsset) - totalDepositsBefore;
+    mathint totalRequestedFundsDiff= totalRequestedFunds(anyAsset) - totalRequestedFundsBefore;
+    mathint totalWithdrawableFundsDiff = totalWithdrawableFunds(anyAsset) - totalWithdrawableFundsBefore;
+    mathint totalWithdrawalsDiff = totalWithdrawals(anyAsset) - totalWithdrawalsBefore;
+
+    assert depositsDiff != 0 => depositsDiff == totalDepositsDiff, "deposits and totalDeposit diff differed";
+    assert requestedFundsDiff != 0 => requestedFundsDiff == totalRequestedFundsDiff, "requestedFunds and totaRequestedFundsdiff differed";
+    assert withdrawableFundsDiff != 0 => withdrawableFundsDiff == totalWithdrawableFundsDiff, "withdrawableFunds and totalWithdrawableFunds diff differed";
+    assert withdrawalsDiff != 0 => withdrawalsDiff == totalWithdrawalsDiff, "withdrawals and totalWithdrawals diff differed";
```
[the remark this answers is not in the document]
Comment (oldchili) [reply certora/ArrangerConduit.spec:838]: If we want that we need to use invariants and ghosts that will some over all ilks.

[the remark this answers is not in the document]
Comment (sunbreak1211) [reply certora/ArrangerConduit.spec:838]: You are right, let's resolve this for now, we will see if it is worth it to do something more complex later.

File: certora/ArrangerConduit.spec [hunk 34]
```diff
@@ -0,0 +1,858 @@
...
+    assert requestAfter.status == IArrangerConduit.StatusEnum.CANCELLED, "cancelFundRequest did not change status to CANCELLED";
+    assert requestAfter.asset == requestBefore.asset
+        && requestAfter.ilk == requestBefore.ilk
+        && requestAfter.amountRequested == requestBefore.amountRequested
+        && requestAfter.amountFilled == requestBefore.amountFilled
+        && infoHashAfter == infoHashBefore,
+        "unrelated request params not as before";
+
+    assert requestedFundsAfter == requestedFundsBefore - requestBefore.amountRequested, "cancelFundRequest did not decrease by amount";
+    assert totalRequestedFundsAfter == totalRequestedFundsBefore - requestBefore.amountRequested, "totalRequestedFunds did not decrease by amount";
```
Comment (telome) [inline certora/ArrangerConduit.spec:597]:
```suggestion
    assert requestedFundsAfter == requestedFundsBefore - requestBefore.amountRequested, "requestedFunds did not decrease by amountRequested";
    assert totalRequestedFundsAfter == totalRequestedFundsBefore - requestBefore.amountRequested, "totalRequestedFunds did not decrease by amountRequested";
```

File: certora/ArrangerConduit.spec (hunk 32 above)
Comment (sunbreak1211) [reply certora/ArrangerConduit.spec:585]: @oldchili let me know what you want to do with this one, if you consider is not worth it will resolve it.

File: certora/ArrangerConduit.spec [hunk 35]
```diff
@@ -593,8 +593,8 @@ rule cancelFundRequest(uint256 fundRequestId) {
         && infoHashAfter == infoHashBefore,
         "unrelated request params not as before";
 
-    assert requestedFundsAfter == requestedFundsBefore - requestBefore.amountRequested, "cancelFundRequest did not decrease by amount";
-    assert totalRequestedFundsAfter == totalRequestedFundsBefore - requestBefore.amountRequested, "totalRequestedFunds did not decrease by amount";
+    assert requestedFundsAfter == requestedFundsBefore - requestBefore.amountRequested, "requestedFunds did not decrease by amountRequested";
+    assert totalRequestedFundsAfter == totalRequestedFundsBefore - requestBefore.amountRequested, "totalRequestedFunds did not decrease by amountRequested";
```
Comment (sunbreak1211) [inline certora/ArrangerConduit.spec:610]: Hmmm I'm a bit confused here, we use to put the name of the function we are evaluating.

**Review by sunbreak1211 — APPROVED** [review]
LGTM

File: certora/ArrangerConduit.spec [hunk 36]
```diff
@@ -0,0 +1,871 @@
...
+    assert lastReverted <=> revert1 || revert2 || revert3 ||
+                            revert4 || revert5 || revert6 ||
+                            revert7, "Revert rules failed";
+}
+
+// Verify correct storage changes for non reverting requestFunds
+rule requestFunds(bytes32 ilk, address asset, uint256 amount, string info) {
+    env e;
+    
+    require info.length % 32 == 0; // See explanation for the in the header on the top
```
Comment (telome) [inline certora/ArrangerConduit.spec:449]:
```suggestion
    require info.length % 32 == 0; // See explanation for that in the header on the top
```

File: certora/ArrangerConduit.spec [hunk 37]
```diff
@@ -0,0 +1,871 @@
...
+        uint256    amountFilled;  // 32 bytes
+        string     info;
+    }
+  * The storage is 32-bytes aligned.
+  * Since the `status` and `asset` fields are packed together the offset is 32 * 4 = 128.
+*/
+
+/** On top of the above, the Certora tool currently can not handle properly hashing of
+*   strings with a length that is not a multiple of 32 (masking out the extra bytes in
+*   the last word). Therefor we also add a `require info.length % 32 == 0`;
```
Comment (telome) [inline certora/ArrangerConduit.spec:84]:
```suggestion
*   the last word). Therefore we also add a `require info.length % 32 == 0`;
```

**Review by telome — APPROVED** [review]
LGTM
